aikido_perception/src/AprilTagsModule.cpp

```diff
@@ -0,0 +1,170 @@
+
+			if(marker_stamp < timestamp){
+				dtwarn << "Marker stamp " << marker_stamp<< " for " << tag_name 
+				<< "is behind timestamp parameter: " <<timestamp <<"\n";
+
+				continue;
+			}
+
+			std::string skel_name;
+			Eigen::Isometry3d skel_offset;
+			dart::common::Uri skel_resource;
+
+			//check if tag name is in database
+			if (mConfigData->getTagNameOffset(tag_name,skel_name,skel_resource,skel_offset))
+			{
+
+				//Get orientation of marker
+				Eigen::Isometry3d marker_pose = aikido::perception::convertROSPoseToEigen(marker_transform.pose);
+
+
+				//For the frame-frame transform
+				tf::StampedTransform transform;
+
+				try{
+					mListener->waitForTransform(mReferenceFrameId,detection_frame,
+						marker_stamp,ros::Duration(_timeout));
+
+					mListener->lookupTransform(mReferenceFrameId,detection_frame,
+						marker_stamp, transform);
+				}
+				catch(const tf::ExtrapolationException& ex){
+					ROS_ERROR("%s",ex.what());
+					continue;
+				}
+				catch(const tf::TransformException& ex){
+					ROS_ERROR("%s",ex.what());
+					throw std::runtime_error("TF Error which cannot be recovered");
+				}
+
+				transform_lookedup = true;
+
+				//Get the transform as a pose matrix
+				Eigen::Isometry3d frame_pose = aikido::perception::convertStampedTransformToEigen(transform);
+
+				//Compose to get actual skeleton pose
+				Eigen::Isometry3d temp_pose = frame_pose * marker_pose;
+				Eigen::Isometry3d skel_pose = temp_pose * skel_offset;
+
+				Eigen::Isometry3d link_offset = mReferenceLink->getWorldTransform();
+				skel_pose = link_offset * skel_pose;
+
+				//Check if skel in skel_list
+				//If there then update pose
+				//If not then append skeleton
+				skel_name.append(std::to_string(marker_transform.id));
+				bool is_new_skel;
+
+				//Search skeleton_list for skeleton
+				const auto it = std::find_if(std::begin(skeleton_list), std::end(skeleton_list),
+					[&](const dart::dynamics::SkeletonPtr& skeleton)
+				  	{
+						return skeleton->getName() == skel_name;
+				  	}
+				);
+
+				dart::dynamics::SkeletonPtr skel_to_update;
+
+				if (it == std::end(skeleton_list)){
+					//New skeleton
+					is_new_skel = true;
+					dart::utils::DartLoader urdfLoader;
+					skel_to_update = 
+						urdfLoader.parseSkeleton(skel_resource,mResourceRetriever);
+					
+					if(!skel_to_update)
+						dtwarn<<"Empty SkeletonPtr for "<<skel_name<<std::endl;
+					skel_to_update->setName(skel_name);
+
+				}
+				else{
+					//Get existing skeletonPtr
+					is_new_skel = false;
+					skel_to_update = *it;
+				}
+
+
+				//Get root joint of skeleton
+				dart::dynamics::Joint* jtptr = skel_to_update->getJoint(0);
+
+				//Downcast Joint to FreeJoint
+				dart::dynamics::FreeJoint* freejtptr = dynamic_cast<dart::dynamics::FreeJoint*>(jtptr);
+
+				//Check if successful down-cast
+				if(freejtptr == NULL){
+					dtwarn << "Could not cast the joint of the body to a Free Joint so ignoring marker "<<tag_name << std::endl;
```

Comment:
The convention in DART is to prefix log output with `[ClassName::functionName]`.